sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala

```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
     DataFrameHolder(
       _sqlContext.internalCreateDataFrame(rows, StructType(StructField("_1", dataType) :: Nil)))
   }
+
+  /**
+   * ::Experimental::
+   *
+   * Pimp my library decorator for tungsten caching of DataFrames.
+   * @since 1.5.1
+   */
+  @Experimental
+  implicit class TungstenCache(df: DataFrame) {
+    /**
+     * Packs the rows of [[df]] into contiguous blocks of memory.
+     * @param compressionType "" (default), "lz4", "lzf", or "snappy", see
+     *                        [[CompressionCodec.ALL_COMPRESSION_CODECS]]
+     * @param blockSize size of each MemoryBlock (default = 4 MB)
+     */
+    def tungstenCache(
+        compressionType: String = "", blockSize: Int = 4000000): (RDD[_], DataFrame) = {
+      val schema = df.schema
+
+      val convert = CatalystTypeConverters.createToCatalystConverter(schema)
+      val internalRows = df.rdd.map(convert(_).asInstanceOf[InternalRow])
```

Comment:
This is going to be moderately inefficient if the input happens to already be in UnsafeRow format, since this will end up performing an extra round of conversions, converting each UnsafeRow into a public row, then turning around and converting it back into an UnsafeRow.

Instead, I think that it would make more sense to write a function which takes an InternalRow and returns an UnsafeRow but which acts as a no-op when applied to existing UnsafeRows.